carsus/io/output/base.py

```diff
@@ -0,0 +1,546 @@
+            "levels_metastable_loggf_threshold":
+            levels_metastable_loggf_threshold,
+            "lines_loggf_threshold": lines_loggf_threshold
+        }
+
+        self.ions = parse_selected_species(ions)
+        self.gfall_reader = gfall_reader
+
+        self.ionization_energies = ionization_energies.base
+        self.ground_levels = ionization_energies.get_ground_levels()
+
+        self.levels_all = self._get_all_levels_data().reset_index()
+        self.lines_all = self._get_all_lines_data(self.levels_all)
+
+        self._create_levels_lines(**self.levels_lines_param)
+        self._create_macro_atom()
+        self._create_macro_atom_references()
+
+    @staticmethod
+    def _create_artificial_fully_ionized(levels):
+        """ Create artificial levels for fully ionized ions """
+        fully_ionized_levels = list()
+
+        for atomic_number, _ in levels.groupby("atomic_number"):
+            fully_ionized_levels.append(
+                (-1, atomic_number, atomic_number, 0, 0.0, 1, True)
+            )
+
+        levels_columns = ["level_id", "atomic_number",
+                          "ion_number", "level_number",
+                          "energy", "g", "metastable"]
+        fully_ionized_levels_dtypes = [
+            (key, levels.dtypes[key]) for key in levels_columns]
+
+        fully_ionized_levels = np.array(
+            fully_ionized_levels, dtype=fully_ionized_levels_dtypes)
+
+        return pd.DataFrame(data=fully_ionized_levels)
+
+    @staticmethod
+    def _create_metastable_flags(levels, lines,
+                                 levels_metastable_loggf_threshold=-3):
+        # Filter lines on the loggf threshold value
+        metastable_lines = lines.loc[lines["loggf"]
+                                     > levels_metastable_loggf_threshold]
+
+        # Count the remaining strong transitions
+        metastable_lines_grouped = metastable_lines.groupby("upper_level_id")
+        metastable_counts = metastable_lines_grouped["upper_level_id"].count()
+        metastable_counts.name = "metastable_counts"
+
+        # If there are no strong transitions for a level (the count is NaN)
+        # then the metastable flag is True else (the count is a natural number)
+        # the metastable flag is False
+        levels = levels.join(metastable_counts)
+        metastable_flags = levels["metastable_counts"].isnull()
+        metastable_flags.name = "metastable"
+
+        return metastable_flags
+
+    def _get_all_levels_data(self):
+        """ Returns the same output than `AtomData._get_all_levels_data()` """
+        gf = self.gfall_reader
+        gf.levels['g'] = 2*gf.levels['j'] + 1
+        gf.levels['g'] = gf.levels['g'].map(np.int)
+
+        ions_df = pd.DataFrame.from_records(
+            self.ions, columns=["atomic_number", "ion_charge"])
+        ions_df = ions_df.set_index(['atomic_number', 'ion_charge'])
+        levels = gf.levels.reset_index().join(ions_df, how="inner",
```

Comment:
I think this might be easier to read if you move the two `reset_index` operations out of this expression.

Reply:
Fixed